File: CSharp/ByRound/1606/1606_E_Arena.cs
// Problem: CF 1606 E - Arena
// https://codeforces.com/contest/1606/problem/E

/*
 * Problem: E. Arena
 * Algorithm: Dynamic Programming with Inclusion-Exclusion Principle
 * Time Complexity: O(n^3)
 * Space Complexity: O(n^2)
 *
 * The problem asks to count the number of ways to assign health points to heroes
 * such that no hero wins (i.e., there's no hero who survives all rounds).
 *
 * Approach:
 * - Use dynamic programming where mas[i, j] represents the number of ways to assign
 *   health points such that exactly i heroes survive and the minimum health is j.
 * - Precompute factorials and binomial coefficients for combinatorial calculations.
 * - Apply inclusion-exclusion principle and dynamic programming to count valid assignments.
 */

using System;
using System.Collections.Generic;
using System.Linq;

namespace VaHiX
{
    class Program
    {
        static long mod = 998244353L;

        static void Main(string[] args)
        {
            var input = Console.ReadLine().Split(' ').Select(int.Parse).ToList();
            int n = input[0],
                x = input[1];

            // Precompute powers of integers up to n
            var xar = new long[n + 1, n + 1];
            for (int i = 1; i <= n; i++)
            {
                xar[i, 0] = 1;
                for (int j = 1; j <= n; j++)
                    xar[i, j] = (xar[i, j - 1] * i) % mod;
            }

            // Precompute binomial coefficients
            var c = new long[n + 1, n + 1];
            c[0, 0] = 1;
            for (int i = 1; i <= n; i++)
            {
                c[i, 0] = 1;
                for (int k = 1; k <= i; k++)
                    c[i, k] = (c[i - 1, k - 1] + c[i - 1, k]) % mod;
            }

            // Dynamic programming calculation
            var mas = new long[n + 1, x + 1];
            for (int i = 1; i <= n; i++)
            {
                for (int j = i; j <= x; j++)
                {
                    if (i == 1)
                    {
                        mas[i, j] = j;
                        continue;
                    }
                    for (int k = 0; k < i; k++)
                        mas[i, j] =
                            (
                                mas[i, j]
                                + (((xar[i - 1, k] * c[i, k]) % mod) * mas[i - k, j - i + 1]) % mod
                            ) % mod;
                }
            }

            // Total number of ways to assign health points (x^n)
            long xn = 1;
            for (int i = 1; i <= n; i++)
                xn = (xn * x) % mod;

            // Result = Total - Ways where at least one hero wins
            var res = (xn - mas[n, x] + mod) % mod;
            Console.Write(res);
        }
    }
}


// https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/E/1606_E_Arena.cpp
// Problem: CF 1606 E - Arena
// https://codeforces.com/contest/1606/problem/E

/*
Algorithm: Dynamic Programming with Combinatorics
Time Complexity: O(n^3)
Space Complexity: O(n^2)

This solution uses dynamic programming with combinatorial mathematics to count the number of ways
to assign initial health points to heroes such that no single hero survives all rounds.

The key insight is to work backwards from the final state where no hero remains alive,
tracking how many heroes could have started with each health value.
We use combinatorics to track combinations and modular arithmetic throughout.
*/

#include <stdio.h>
#include <algorithm>
#include <iostream>

#define LL long long
#define mod 998244353
using namespace std;
const int N = 509;
int n, x;
LL ans;
LL c[N][N], dp[N][N];

int main() {
  cin >> n >> x;
  // Precompute binomial coefficients C(i,j) for all i,j up to n
  for (int i = 0; i <= n; i++) {
    c[i][0] = c[i][i] = 1ll;
    for (int j = 1; j < i; j++)
      c[i][j] = (c[i - 1][j - 1] + c[i - 1][j]) % mod;
  }
  
  // Initialize base case: 1 way to have 0 heroes alive with 0 health points
  dp[n][0] = 1ll;
  
  // Work backwards from n heroes to 1 hero (eliminating one hero each round)
  for (int i = n; i > 1; i--)
    for (int j = 0; j < x; j++) {
      // Calculate the new maximum health value for the next round
      int nj = min(x, j + i - 1);
      LL pw = 1ll;
      // For each possible number of remaining heroes (k) in the next round
      for (int k = i; k >= 0; k--) {
        // Update dp[k][nj] by adding contribution from dp[i][j]
        // Using combinatorial coefficient C(i,k) and power term (nj-j)^k
        dp[k][nj] = (dp[k][nj] + dp[i][j] * (c[i][k] * pw % mod) % mod) % mod;
        pw = pw * (LL)(nj - j) % mod;
      }
    }
  
  // Sum up all possible configurations where no hero wins (all have 0 heroes left)
  for (int i = 0; i <= x; i++)
    ans = (ans + dp[0][i]) % mod;
  
  printf("%lld\n", ans);
  return 0;
}


// https://github.com/VaHiX/CodeForces/